File: bindings/deno/src/safe_hex.ts
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/**
 * SafeHex - Hexadecimal encoding/decoding.
 *
 * Thin FFI wrapper: all computation delegates to libproven (Idris 2 + Zig).
 *
 * @module
 */

import { getLib, ProvenStatus, readAndFreeString, statusToError } from './ffi.ts';
import { err, ok, type Result } from './result.ts';

/**
 * Safe hex operations backed by formally verified Idris 2 code.
 */
export class SafeHex {
  /**
   * Encode bytes to hexadecimal string.
   *
   * @param bytes - The bytes to encode.
   * @param uppercase - Whether to use uppercase hex digits.
   * @returns Result containing the hex string or an error.
   */
  static encode(bytes: Uint8Array, uppercase = false): Result<string> {
    const symbols = getLib();
    const result = symbols.proven_hex_encode(bytes, bytes.length, uppercase);
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    const encoded = readAndFreeString(result[1], Number(result[2]));
    if (encoded === null) return err('Null string returned');
    return ok(encoded);
  }
}

File: bindings/java/src/main/java/io/github/hyperpolymath/proven/SafeHex.java
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
package io.github.hyperpolymath.proven;

import com.sun.jna.Memory;

import java.util.Optional;

/**
 * Safe hexadecimal encoding via libproven FFI.
 *
 * All logic delegates to the Idris 2 verified implementation through
 * the Zig C ABI. No hex encoding logic is reimplemented in Java.
 */
public final class SafeHex {

    private static final ProvenLibrary LIB = ProvenLibrary.INSTANCE;

    private SafeHex() {}

    /**
     * Encode bytes to lowercase hex string.
     *
     * @param data the bytes to encode
     * @return the hex string, or empty on error
     */
    public static Optional<String> encode(byte[] data) {
        return encodeInternal(data, false);
    }

    /**
     * Encode bytes to uppercase hex string.
     *
     * @param data the bytes to encode
     * @return the hex string, or empty on error
     */
    public static Optional<String> encodeUpper(byte[] data) {
        return encodeInternal(data, true);
    }

    private static Optional<String> encodeInternal(byte[] data, boolean uppercase) {
        Memory mem = NativeUtil.toNativeBytes(data);
        if (mem == null) {
            return Optional.empty();
        }
        return NativeUtil.extractString(
            LIB.proven_hex_encode(mem, data.length, uppercase));
    }
}
